### PR TITLE
fix `xexpy(0, 1000)`

### PR DESCRIPTION
**Pull request by oscardssmith**
now it equals 0.

### CODE REVIEW

**Comment by oscardssmith** [comment]
tests added.

**Comment by devmotion** [comment]
The PR does not match what `xexpy` is intended to return: https://github.com/JuliaStats/LogExpFunctions.jl/issues/79#issuecomment-1944223377

**Comment by oscardssmith** [comment]
@devmotion I don't understand what you mean?

**Comment by devmotion** [comment]
`xexpy(0, Inf)` is supposed to (as documented in the docstring) return `0 * exp(Inf)` which is `NaN`. To me it seems there is no bug to fix here.

To elaborate a bit more, `xexpy` and `xlogy` are supposed to satisfy the invariant `xexpy(a, b) = xlogy(exp(b), exp(a))`: If `b = -Inf`, or equivalently `exp(b) = 0`, then `xexpy(a, b) = xlogy(exp(b), exp(a)) = 0`; otherwise they return `a * exp(b) = exp(b) * log(exp(a))`.

**Comment by oscardssmith** [comment]
This PR maintains `xexpy(0, Inf) == NaN`. The difference is in `xexpy(0, 10000)` where previously `exp(10000)` overflowed.

**Comment by devmotion** [comment]
> Why? This requirement is not documented.

The behaviour of each function is documented which implicitly documents the invariance.

> I would also note that xlogy(1e100, 1) == 0, while xlogy(exp(1000), exp(0)) gives NaN only because the intermediate exp(1000) overflows.

My comment was supposed to be understood as a exact/mathematical description of the intended functions and their relation, disregarding floating-point issues such as overflow. That is, the invariance is for the mathematical functions underlying `xlogy` and `xexpy`,  but not their Julia implementations.

**Comment by oscardssmith** [comment]
In that case, isn't this PR correct? This PR leaves xexpy(0, Inf) as NaN.

**Review by devmotion — COMMENTED** [review]
I missed the check for `isfinite(y)`, so I think generally the PR is fine. It still disagrees with the docstring though since due to the change `xexpy(x, y)` is not always `x * exp(y)` if `y != -Inf` anymore. Can you adjust the docstring (e.g., by changing ``if `y == -Inf` `` to ``if `y == -Inf` or if `x == 0.0` and `y` is finite.``) and maybe also add the `xexpy(0, 1000)` example as a doctest?

**Comment by oscardssmith** [comment]
I interpret the docstring as describing the function mathematically, in which case it is already correct.

**Comment by devmotion** [comment]
Generally, the docstrings in LogExpFunctions mention explicitly if the computation deviates from the given Julia code ("preserving numerical accuracy", "carefully evaluated", "avoiding over/underflow" ...), so IMO the docstring should either mention the additional special case or contain a comment that indicates that the computation is not just `x * exp(y)`.

**Comment by oscardssmith** [comment]
Ideally this code could avoid more overflow to deal with cases like `xexpy(floatmin(Float64), 1000.0)`, but that is rather difficult.

File: src/basicfuns.jl [hunk 1]
```diff
@@ -62,7 +62,7 @@ end
 """
 $(SIGNATURES)
 
-Return `x * exp(y)` for `y > -Inf`, or zero if `y == -Inf`.
+Return `x * exp(y)` for `y > -Inf`, or zero if `y == -Inf` or x == 0 and y is finite.
```
Comment (devmotion) [inline src/basicfuns.jl:65]: To improve the display in the docs:
```suggestion
Return `x * exp(y)` for `y > -Inf`, or zero if `y == -Inf` or if `x == 0` and `y` is finite.
```

**Review by devmotion — APPROVED** [review]

**Review by tpapp — APPROVED** [review]

**Pull request merged**